### PR TITLE
[rush-amazon-s3-build-cache-plugin] Read credentials from common AWS_* env vars

### LINKED ISSUE #4115: INFORMATION REQUEST: [rush-amazon-s3-build-cache-plugin] Does build-cache with AWS S3 Support InstanceRoles
Asked for this information in the zulip chat, did not receive an answer so asking here:

## Summary
Interested in finding out if I need to specify Access/SecretKey for AWS S3 usage with build-cache or if it will be able to work if I run without specifying creds on a server that has an IAM role attached to it.

### PR DESCRIPTION
**Pull request by john-u**
## Summary

- check for existence of commonly used environment variables `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN`
- add unit tests for new function as well as existing function for `RUSH_BUILD_CACHE_CREDENTIAL`
- resolve #4115
## Details

Added a function for reading `AWS` vars that easily falls back to the Rush build cache env var if they aren't present. Refactored existing credential parsing (extracted out to module) to aid in unit testing. I considered importing https://github.com/aws/aws-sdk-js-v3/tree/main/packages/credential-provider-node but could not figure out build issues once added, it's possible they don't support TypeScript 5 yet. The `fromEnv` provider they export is essentially what I added here and it a trivial amount of code.
## How it was tested

1. Assumed AWS IAM role using `aws sts assume-role` that has access to build cache bucket
1. `eval $(aws configure export-credentials --format env)` to set cred vars
1.  Invoked `rush build` with `"isCacheWriteAllowed": true` and verified that the build cache was successfully read/updated locally and in s3 using the credentials exported.

## Impacted documentation

https://rushjs.io/pages/maintainer/build_cache/#ci-setup 
While these changes did not modify documented behavior of existing rush env var, we could call out that it checks for `AWS_*` vars for `amazon-s3` now.

### CODE REVIEW

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3Credentials.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import { EnvironmentConfiguration } from '@rushstack/rush-sdk';
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3Credentials.ts:1]:
```suggestion
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { EnvironmentConfiguration } from '@rushstack/rush-sdk';
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/index.ts [hunk 2]
```diff
@@ -3,7 +3,8 @@
 
 import { RushAmazonS3BuildCachePlugin } from './RushAmazonS3BuildCachePlugin';
 
-export { AmazonS3Client, IAmazonS3Credentials } from './AmazonS3Client';
+export { IAmazonS3Credentials } from './AmazonS3Credentials';
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/index.ts:6]:
```suggestion
export { type IAmazonS3Credentials } from './AmazonS3Credentials';
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Client.test.ts [hunk 3]
```diff
@@ -5,8 +5,9 @@ import { ConsoleTerminalProvider, Terminal } from '@rushstack/node-core-library'
 import { Response, ResponseInit } from 'node-fetch';
 
 import { IAmazonS3BuildCacheProviderOptionsAdvanced } from '../AmazonS3BuildCacheProvider';
-import { AmazonS3Client, IAmazonS3Credentials } from '../AmazonS3Client';
+import { AmazonS3Client } from '../AmazonS3Client';
 import { WebClient } from '../WebClient';
+import { IAmazonS3Credentials } from '../AmazonS3Credentials';
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Client.test.ts:10]:
```suggestion
import type { IAmazonS3Credentials } from '../AmazonS3Credentials';
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3Client.ts [hunk 4]
```diff
@@ -7,23 +7,13 @@ import * as fetch from 'node-fetch';
 
 import { IAmazonS3BuildCacheProviderOptionsAdvanced } from './AmazonS3BuildCacheProvider';
 import { IGetFetchOptions, IPutFetchOptions, WebClient } from './WebClient';
+import { IAmazonS3Credentials, fromRushEnv } from './AmazonS3Credentials';
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3Client.ts:10]:
```suggestion
import { type IAmazonS3Credentials, fromRushEnv } from './AmazonS3Credentials';
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3BuildCacheProvider.ts [hunk 5]
```diff
@@ -12,8 +12,9 @@ import {
   EnvironmentConfiguration
 } from '@rushstack/rush-sdk';
 
-import { AmazonS3Client, IAmazonS3Credentials } from './AmazonS3Client';
+import { AmazonS3Client } from './AmazonS3Client';
 import { WebClient } from './WebClient';
+import { IAmazonS3Credentials, fromAmazonEnv, fromRushEnv } from './AmazonS3Credentials';
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3BuildCacheProvider.ts:17]:
```suggestion
import { type IAmazonS3Credentials, fromAmazonEnv, fromRushEnv } from './AmazonS3Credentials';
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
+import {
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts:1]:
```suggestion
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import {
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+      const credentials = fromAmazonEnv();
+
+      expect(credentials).toBeDefined();
+      expect(credentials!.accessKeyId).toBe(AWS_ACCESS_KEY_ID);
+      expect(credentials!.secretAccessKey).toBe(AWS_SECRET_ACCESS_KEY);
+      expect(credentials!.sessionToken).toBeUndefined();
+    });
+
+    it('returns undefined if access key and secret are not both present', () => {
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts:43]: Would we want the plugin to throw instead if the env variables don't provide a complete credential?

File: common/changes/@microsoft/rush/s3-build-cache-read-aws-creds_2023-06-02-00-43.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "[rush-amazon-s3-build-cache-plugin] Read credentials from common AWS_* env vars.",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/s3-build-cache-read-aws-creds_2023-06-02-00-43.json:5]:
```suggestion
      "comment": "Add functionality for the Amazon S3 Build Cache Plugin to read credentials from common AWS_* environment variables.",
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+} from '../AmazonS3Credentials';
+import { EnvironmentConfiguration } from '@rushstack/rush-sdk';
+
+describe('Amazon S3 Credentials', () => {
+  describe(fromAmazonEnv.name, () => {
+    afterEach(() => {
+      delete process.env[AWS_ACCESS_KEY_ID];
+      delete process.env[AWS_SECRET_ACCESS_KEY];
+      delete process.env[AWS_SESSION_TOKEN];
+    });
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts:16]: Can you back up any existing values these might have first and restore them afterward so `process.env` can't be permanently modified by this test?

**Review by iclanton — APPROVED** [review]

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts:19]:
```suggestion
    let oldEnvAwsAccessKeyId: string | undefined;
    let oldEnvAwsSecretAccessKey: string | undefined;
    let oldEnvAwsSessionToken: string | undefined;
    
    beforeEach(() => {
      oldEnvAwsAccessKeyId = process.env[AWS_ACCESS_KEY_ID];
      oldEnvAwsSecretAccessKey = process.env[AWS_SECRET_ACCESS_KEY];
      oldEnvAwsSessionToken = process.env[AWS_SESSION_TOKEN];
    });
    
    afterEach(() => {
      process.env[AWS_ACCESS_KEY_ID] = oldEnvAwsAccessKeyId;
      process.env[AWS_SECRET_ACCESS_KEY] = oldEnvAwsSecretAccessKey;
      process.env[AWS_SESSION_TOKEN] = oldEnvAwsSessionToken;
    });
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3Credentials.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+  if (accessKeyId && secretAccessKey) {
+    return {
+      accessKeyId,
+      secretAccessKey,
+      sessionToken
+    };
+  }
+
+  return undefined;
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3Credentials.ts:37]:
```suggestion
  if (accessKeyId && secretAccessKey) {
    return {
      accessKeyId,
      secretAccessKey,
      sessionToken
    };
  } else if (accessKeyId) {
    throw new Error(
      `The "${AWS_ACCESS_KEY_ID}" env variable is set, but the "${AWS_SECRET_ACCESS_KEY}" `+
      `env variable is not set. Both or neither must be provided.`
    );
  } else if (secretAccessKey) {
    throw new Error(
      `The "${AWS_SECRET_ACCESS_KEY}" env variable is set, but the "${AWS_ACCESS_KEY_ID}" `+
      `env variable is not set. Both or neither must be provided.`
    );
  } else {
    return undefined;
  }
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+  describe(fromAmazonEnv.name, () => {
+    let oldEnvAwsAccessKeyId: string | undefined;
+    let oldEnvAwsSecretAccessKey: string | undefined;
+    let oldEnvAwsSessionToken: string | undefined;
+    
+    beforeEach(() => {
+      oldEnvAwsAccessKeyId = process.env[AWS_ACCESS_KEY_ID];
+      oldEnvAwsSecretAccessKey = process.env[AWS_SECRET_ACCESS_KEY];
+      oldEnvAwsSessionToken = process.env[AWS_SESSION_TOKEN];
+    });
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts:23]:
```suggestion
    beforeEach(() => {
      oldEnvAwsAccessKeyId = process.env[AWS_ACCESS_KEY_ID];
      oldEnvAwsSecretAccessKey = process.env[AWS_SECRET_ACCESS_KEY];
      oldEnvAwsSessionToken = process.env[AWS_SESSION_TOKEN];
      
      delete process.env[AWS_ACCESS_KEY_ID];
      delete process.env[AWS_SECRET_ACCESS_KEY];
      delete process.env[AWS_SESSION_TOKEN];
    });
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+      expect(credentials!.secretAccessKey).toBe(AWS_SECRET_ACCESS_KEY);
+      expect(credentials!.sessionToken).toBeUndefined();
+    });
+
+    it('returns undefined if access key and secret are not both present', () => {
+      process.env[AWS_ACCESS_KEY_ID] = AWS_ACCESS_KEY_ID;
+
+      const credentials = fromAmazonEnv();
+
+      expect(credentials).toBeUndefined();
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts:61]:
```suggestion
      process.env[AWS_ACCESS_KEY_ID] = AWS_ACCESS_KEY_ID;
      expect(() => fromAmazonEnv()).toThrowErrorMatchingSnapshot();

      delete process.env[AWS_ACCESS_KEY_ID];
      process.env[AWS_SECRET_ACCESS_KEY] = AWS_SECRET_ACCESS_KEY;
      expect(() => fromAmazonEnv()).toThrowErrorMatchingSnapshot();
```

**Comment by Faithfinder** [comment]
Can I quickly confirm that RUSH_BUILD_CACHE_CREDENTIAL has priority over AWS vars? Otherwise it's quite a breaking change

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+      delete process.env[AWS_ACCESS_KEY_ID];
+      delete process.env[AWS_SECRET_ACCESS_KEY];
+      delete process.env[AWS_SESSION_TOKEN];
+    });
+    
+    afterEach(() => {
+      process.env[AWS_ACCESS_KEY_ID] = oldEnvAwsAccessKeyId;
+      process.env[AWS_SECRET_ACCESS_KEY] = oldEnvAwsSecretAccessKey;
+      process.env[AWS_SESSION_TOKEN] = oldEnvAwsSessionToken;
+    });
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts:33]:
```suggestion
    let isOldEnvAwsAccessKeyIdSet: boolean;
    let oldEnvAwsAccessKeyId: string | undefined;
    let isOldEnvAwsSecretAccessKeySet: boolean;
    let oldEnvAwsSecretAccessKey: string | undefined;
    let isOldEnvAwsSessionTokenSet: boolean;
    let oldEnvAwsSessionToken: string | undefined;

    beforeEach(() => {
      isOldEnvAwsAccessKeyIdSet = AWS_ACCESS_KEY_ID in process.env;
      oldEnvAwsAccessKeyId = process.env[AWS_ACCESS_KEY_ID];
      isOldEnvAwsSecretAccessKeySet = AWS_SECRET_ACCESS_KEY in process.env;
      oldEnvAwsSecretAccessKey = process.env[AWS_SECRET_ACCESS_KEY];
      isOldEnvAwsSessionTokenSet = AWS_SESSION_TOKEN in process.env;
      oldEnvAwsSessionToken = process.env[AWS_SESSION_TOKEN];

      delete process.env[AWS_ACCESS_KEY_ID];
      delete process.env[AWS_SECRET_ACCESS_KEY];
      delete process.env[AWS_SESSION_TOKEN];
    });

    afterEach(() => {
      if (isOldEnvAwsAccessKeyIdSet) {
        process.env[AWS_ACCESS_KEY_ID] = oldEnvAwsAccessKeyId;
      } else {
        delete process.env[AWS_ACCESS_KEY_ID];
      }

      if (isOldEnvAwsSecretAccessKeySet) {
        process.env[AWS_SECRET_ACCESS_KEY] = oldEnvAwsSecretAccessKey;
      } else {
        delete process.env[AWS_SECRET_ACCESS_KEY];
      }

      if (isOldEnvAwsSessionTokenSet) {
        process.env[AWS_SESSION_TOKEN] = oldEnvAwsSessionToken;
      } else {
        delete process.env[AWS_SESSION_TOKEN];
      }
    });
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+      expect(credentials!.sessionToken).toBeUndefined();
+    });
+
+    it('returns undefined if access key and secret are not both present', () => {
+      process.env[AWS_ACCESS_KEY_ID] = AWS_ACCESS_KEY_ID;
+      expect(() => fromAmazonEnv()).toThrowErrorMatchingSnapshot();
+
+      delete process.env[AWS_ACCESS_KEY_ID];
+      process.env[AWS_SECRET_ACCESS_KEY] = AWS_SECRET_ACCESS_KEY;
+      expect(() => fromAmazonEnv()).toThrowErrorMatchingSnapshot();
```
Comment (iclanton) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/test/AmazonS3Credentials.test.ts:66]:
```suggestion
      expect(() => fromAmazonEnv()).toThrowErrorMatchingInlineSnapshot(`"The \\"AWS_ACCESS_KEY_ID\\" env variable is set, but the \\"AWS_SECRET_ACCESS_KEY\\" env variable is not set. Both or neither must be provided."`);

      delete process.env[AWS_ACCESS_KEY_ID];
      process.env[AWS_SECRET_ACCESS_KEY] = AWS_SECRET_ACCESS_KEY;
      expect(() => fromAmazonEnv()).toThrowErrorMatchingInlineSnapshot(`"The \\"AWS_SECRET_ACCESS_KEY\\" env variable is set, but the \\"AWS_ACCESS_KEY_ID\\" env variable is not set. Both or neither must be provided."`);
```

**Comment by Faithfinder** [comment]
> @Faithfinder - It currently doesn't, but that's a simple change.

I believe it should. Our cache is stored in a separate AWS account from the production environment. If AWS vars would take precedence, we'd lose cache when building for prod, and I guess it would actually be non-trivial to work around?

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3BuildCacheProvider.ts [hunk 16]
```diff
@@ -99,11 +98,11 @@ export class AmazonS3BuildCacheProvider implements ICloudBuildCacheProvider {
 
   private async _getS3ClientAsync(terminal: ITerminal): Promise<AmazonS3Client> {
     if (!this.__s3Client) {
-      let credentials: IAmazonS3Credentials | undefined = AmazonS3Client.tryDeserializeCredentials(
-        this._environmentCredential
-      );
+      let credentials: IAmazonS3Credentials | undefined = fromAmazonEnv() ?? fromRushEnv();
```
[the remark this answers is not in the document]
Comment (iclanton) [reply rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3BuildCacheProvider.ts:101]: @Faithfinder - This will restore the existing behavior.

**Comment by iclanton** [comment]
@john-u - Can you sign the CLA?

**Comment by john-u** [comment]
@microsoft-github-policy-service agree

**Pull request merged**